File: ts/src/source.ts
// Token sources — pull a raw (still-encoded) token out of a request's headers.
//
// A TokenSource is a plain function so composing them (chain) and adapting a
// framework needs no classes. The header names below are the ones the fleet's
// gateway (oauth2-proxy) sets, plus Authorization; every one is overridable.

import type { Headers } from "./identity.js";

/**
 * HEADER carries the header names oauth2-proxy sets, plus Authorization.
 * Lookups are case-insensitive, so the canonical casing here is cosmetic.
 */
export const HEADER = {
  /**
   * accessToken carries the verified access token oauth2-proxy forwards
   * (set_xauthrequest + pass_access_token). Not "Bearer"-prefixed.
   */
  accessToken: "x-auth-request-access-token",
  /** user / email / groups are oauth2-proxy's pre-parsed identity headers — trust mode only. */
  user: "x-auth-request-user",
  email: "x-auth-request-email",
  groups: "x-auth-request-groups",
  /** authorization is the standard bearer header, for direct API callers and tests. */
  authorization: "authorization",
} as const;

/** TokenSource pulls a raw (still-encoded) token out of a request, or undefined. */
export type TokenSource = (headers: Headers) => string | undefined;

/**
 * headerSource reads a token from one request header.
 *
 * `scheme`, if set, is stripped case-insensitively together with a following
 * space: scheme "Bearer" turns "Bearer abc" into "abc" (RFC 7235 makes the
 * scheme case-insensitive, and proxies do vary it). Omitting `scheme` reads the
 * header verbatim — how oauth2-proxy sets x-auth-request-access-token.
 */
export function headerSource(name: string, scheme?: string): TokenSource {
  return (headers) => {
    const value = (headers.get(name) ?? "").trim();
    if (value === "") {
      return undefined;
    }

    if (!scheme) {
      return value;
    }

    const prefix = scheme + " ";
    if (value.length < prefix.length || value.slice(0, prefix.length).toLowerCase() !== prefix.toLowerCase()) {
      return undefined;
    }

    const token = value.slice(prefix.length).trim();

    return token === "" ? undefined : token;
  };
}

/** chain tries each source in order and returns the first token found. */
export function chain(...sources: TokenSource[]): TokenSource {
  return (headers) => {
    for (const source of sources) {
      const raw = source(headers);
      if (raw !== undefined) {
        return raw;
      }
    }

    return undefined;
  };
}

/** oauth2ProxyAccessToken reads the access token oauth2-proxy injects. */
export function oauth2ProxyAccessToken(): TokenSource {
  return headerSource(HEADER.accessToken);
}

/** authorizationBearer reads a bearer token from the Authorization header. */
export function authorizationBearer(): TokenSource {
  return headerSource(HEADER.authorization, "Bearer");
}

/**
 * defaultSource tries the oauth2-proxy header first, then Authorization —
 * covering both a gateway-fronted request and a direct API caller.
 */
export function defaultSource(): TokenSource {
  return chain(oauth2ProxyAccessToken(), authorizationBearer());
}

/**
 * splitAndTrim splits on sep and drops empty/whitespace entries — for the
 * comma-separated groups header.
 */
export function splitAndTrim(value: string, sep: string): string[] {
  return value
    .split(sep)
    .map((part) => part.trim())
    .filter((part) => part !== "");
}

File: source.go
package gatewayauth

import "strings"

// Header names the fleet's gateway (oauth2-proxy) sets, plus Authorization.
const (
	// HeaderAccessToken carries the verified access token oauth2-proxy
	// forwards (set_xauthrequest + pass_access_token). Not "Bearer"-prefixed.
	HeaderAccessToken = "X-Auth-Request-Access-Token" //nolint:gosec // G101 false positive: an HTTP header NAME, not a credential.
	// HeaderUser / HeaderEmail / HeaderGroups are oauth2-proxy's pre-parsed
	// identity headers — used by the trust-mode authenticator only.
	HeaderUser   = "X-Auth-Request-User"
	HeaderEmail  = "X-Auth-Request-Email"
	HeaderGroups = "X-Auth-Request-Groups"
	// HeaderAuthorization is the standard bearer header, for direct API
	// callers and tests.
	HeaderAuthorization = "Authorization"
)

// TokenSource pulls a raw (still-encoded) token out of a request.
type TokenSource interface {
	Token(h Headers) (raw string, ok bool)
}

// HeaderSource reads a token from one request header.
//
// Scheme, if set, is stripped case-insensitively together with a following
// space: Scheme "Bearer" turns "Bearer abc" into "abc" (RFC 7235 makes the
// scheme case-insensitive, and proxies do vary it). Scheme "" reads the header
// verbatim — how oauth2-proxy sets X-Auth-Request-Access-Token.
type HeaderSource struct {
	Name   string
	Scheme string
}

// Token implements TokenSource.
func (s HeaderSource) Token(h Headers) (string, bool) {
	value := strings.TrimSpace(h.Get(s.Name))
	if value == "" {
		return "", false
	}

	if s.Scheme == "" {
		return value, true
	}

	prefix := s.Scheme + " "
	if len(value) < len(prefix) || !strings.EqualFold(value[:len(prefix)], prefix) {
		return "", false
	}

	token := strings.TrimSpace(value[len(prefix):])

	return token, token != ""
}

// ChainSource tries each source in order and returns the first token found.
type ChainSource []TokenSource

// Token implements TokenSource.
func (c ChainSource) Token(h Headers) (string, bool) {
	for _, source := range c {
		if raw, ok := source.Token(h); ok {
			return raw, true
		}
	}

	return "", false
}

// OAuth2ProxyAccessToken reads the access token oauth2-proxy injects.
func OAuth2ProxyAccessToken() TokenSource { return HeaderSource{Name: HeaderAccessToken} }

// AuthorizationBearer reads a bearer token from the Authorization header.
func AuthorizationBearer() TokenSource {
	return HeaderSource{Name: HeaderAuthorization, Scheme: "Bearer"}
}

// DefaultSource tries the oauth2-proxy header first, then Authorization —
// covering both a gateway-fronted request and a direct API caller.
func DefaultSource() TokenSource {
	return ChainSource{OAuth2ProxyAccessToken(), AuthorizationBearer()}
}

// splitAndTrim splits on sep and drops empty/whitespace entries — for the
// comma-separated groups header.
func splitAndTrim(s, sep string) []string {
	parts := strings.Split(s, sep)
	out := make([]string, 0, len(parts))

	for _, part := range parts {
		if trimmed := strings.TrimSpace(part); trimmed != "" {
			out = append(out, trimmed)
		}
	}

	if len(out) == 0 {
		return nil
	}

	return out
}

// ForwardAuthorization renders the caller's credential as an Authorization
// header value for a downstream call made ON BEHALF of the caller (a console
// forwarding the operator's token to its broker, say). It prefers a verbatim
// Authorization header and falls back to the gateway's forwarded access token
// (adding the Bearer scheme oauth2-proxy strips). Empty when the request
// carries no credential — the downstream service then refuses it, which is
// the honest outcome.
func ForwardAuthorization(h Headers) string {
	if a := strings.TrimSpace(h.Get(HeaderAuthorization)); a != "" {
		return a
	}

	if t := strings.TrimSpace(h.Get(HeaderAccessToken)); t != "" {
		return "Bearer " + t
	}

	return ""
}

// HeaderGetter adapts a plain header-lookup function (http.Header.Get,
// connect req.Header().Get) to the Headers interface.
type HeaderGetter func(name string) string

// Get satisfies Headers.
func (g HeaderGetter) Get(name string) string { return g(name) }
